Enable BFCache by default in WebKitTestRunner for Navigation API tests

RupinMittal opened the pull request:
#### f1e512aef1b67e097379833c333ea6c975450f66
<pre>
Enable BFCache by default in WebKitTestRunner for Navigation API tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=299917">https://bugs.webkit.org/show_bug.cgi?id=299917</a>
<a href="https://rdar.apple.com/161693746">rdar://161693746</a>

Reviewed by Chris Dumez.

Some WPT Navigation API tests require the BFCache. This is off by default for
WKTR which means we can&apos;t run these tests. 300439@main tried to skip the WPT
tests and make a copy of them in the http/wpt folder, but these tests were
skipped in 300617@main. So these tests aren&apos;t being tested at all.

To allow these tests to run and therefore improve our test coverage, we enable
the BFCache by default in WKTR for Navigation API tests.

We also remove the copied tests created in 300439@main.

With this change, these two tests are timeouts (as they are on WPT). They were
passing locally before, but are failing on WPT. Since that&apos;s our source of truth,
we mark them as timeouts here as well and will investigate in a subsequent patch:
cross-origin-traversal-does-not-fire-navigate.html
navigation-back-cross-document-preventDefault.html

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/navigation-api/navigate-event/navigate-history-back-bfcache-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/navigate-event/navigate-history-back-bfcache.html: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-activation/activation-after-bfcache-cross-origin-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-activation/activation-after-bfcache-cross-origin.html: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-activation/activation-after-bfcache-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-activation/activation-after-bfcache.html: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-history-entry/entries-after-bfcache-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe.html: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-history-entry/entries-after-bfcache.html: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-history-entry/resources/is_uuid.js: Removed.
* LayoutTests/http/wpt/navigation-api/per-entry-events/dispose-after-bfcache-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/per-entry-events/dispose-after-bfcache.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-origin-traversal-does-not-fire-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-back-cross-document-preventDefault-expected.txt:
* LayoutTests/platform/wk2/TestExpectations:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageClearBackForwardCache):
* Source/WebKit/UIProcess/API/C/WKPage.h:
* Tools/TestRunnerShared/TestFeatures.cpp:
(WTR::shouldUseBackForwardCache):
(WTR::hardcodedFeaturesBasedOnPathForTest):

If the test path contains &quot;navigation-api/&quot;, use the BFCache.

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::generatePageConfiguration):

If the TestOptions flag for UsesBackForwardCache is true, set the
UsesBackForwardCache flag on the Preferences object to be true, so
that the Page knows to use the BFCache.

(WTR::TestController::resetStateToConsistentValues):

Clear the BFCache for the Test Page in between tests

* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::usesBackForwardCache const):

Canonical link: <a href="https://commits.webkit.org/300920@main">https://commits.webkit.org/300920@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fd4ce7ad0edd28f9ec5e8a3d11be4ef7763b49c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124269 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43955 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34679 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131105 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76341 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/591d81ff-7603-4dca-97b8-726c7fe3c171) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126146 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44698 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52557 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94527 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62721 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127223 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35619 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111161 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75115 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d44a9dbd-86f3-4720-9dc8-25ffb356564f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34564 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29320 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74586 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105363 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29542 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133776 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51182 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39022 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103006 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51591 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107377 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102804 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48177 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26425 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48122 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19524 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51044 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56826 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50483 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53839 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52158 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b72c85991a4d13704777917c3d1c0e30dcda544b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51663

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1fd4ce7ad0edd28f9ec5e8a3d11be4ef7763b49c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51663

RupinMittal commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -1445,6 +1449,9 @@ bool TestController::resetStateToConsistentValues(const TestOptions& options, Re
     WKURLRequestSetDefaultTimeoutInterval((60_s).value());
 
+    WKRetainPtr configuration = adoptWK(WKPageCopyPageConfiguration(mainWebView()->page()));
+    WKPageConfigurationClearBackForwardCache(configuration.get());
```
> Why isn't the new API on WKPage since we want to clear bf items for a given WKWebView?

Just moved it to `WKPage`. 

> Also, we probably want to call removeEntriesForPage() on the WebBackForwardCache instead of clear() to only clear the items for the WebView in question no?

I think you're right. For tests that create other WebViews, they should also destroy those WebViews at the end (which clears the entries for those WebViews). So we shouldn't need to clear those manually.

As for the `BackForwardCache`, it looks like whenever a `WebBackForwardCacheEntry` is destroyed (which is what we're doing with the new API function), it clears the corresponding entry in `BackForwardCache`, so we shouldn't need to clear those manually.

Let me know if that sounds right.

cdumez reviewed (commented):
LGTM if the bots are happy.

RupinMittal committed:
Enable BFCache by default in WebKitTestRunner for Navigation API tests
https://bugs.webkit.org/show_bug.cgi?id=299917
rdar://161693746

Reviewed by Chris Dumez.

Some WPT Navigation API tests require the BFCache. This is off by default for
WKTR which means we can't run these tests. 300439@main tried to skip the WPT
tests and make a copy of them in the http/wpt folder, but these tests were
skipped in 300617@main. So these tests aren't being tested at all.

To allow these tests to run and therefore improve our test coverage, we enable
the BFCache by default in WKTR for Navigation API tests.

We also remove the copied tests created in 300439@main.

With this change, these two tests are timeouts (as they are on WPT). They were
passing locally before, but are failing on WPT. Since that's our source of truth,
we mark them as timeouts here as well and will investigate in a subsequent patch:
cross-origin-traversal-does-not-fire-navigate.html
navigation-back-cross-document-preventDefault.html

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/navigation-api/navigate-event/navigate-history-back-bfcache-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/navigate-event/navigate-history-back-bfcache.html: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-activation/activation-after-bfcache-cross-origin-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-activation/activation-after-bfcache-cross-origin.html: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-activation/activation-after-bfcache-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-activation/activation-after-bfcache.html: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-history-entry/entries-after-bfcache-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe.html: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-history-entry/entries-after-bfcache.html: Removed.
* LayoutTests/http/wpt/navigation-api/navigation-history-entry/resources/is_uuid.js: Removed.
* LayoutTests/http/wpt/navigation-api/per-entry-events/dispose-after-bfcache-expected.txt: Removed.
* LayoutTests/http/wpt/navigation-api/per-entry-events/dispose-after-bfcache.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-origin-traversal-does-not-fire-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-back-cross-document-preventDefault-expected.txt:
* LayoutTests/platform/wk2/TestExpectations:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageClearBackForwardCache):
* Source/WebKit/UIProcess/API/C/WKPage.h:
* Tools/TestRunnerShared/TestFeatures.cpp:
(WTR::shouldUseBackForwardCache):
(WTR::hardcodedFeaturesBasedOnPathForTest):

If the test path contains "navigation-api/", use the BFCache.

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::generatePageConfiguration):

If the TestOptions flag for UsesBackForwardCache is true, set the
UsesBackForwardCache flag on the Preferences object to be true, so
that the Page knows to use the BFCache.

(WTR::TestController::resetStateToConsistentValues):

Clear the BFCache for the Test Page in between tests

* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::usesBackForwardCache const):

Canonical link: https://commits.webkit.org/300920@main
